### DISCUSSION TITLE
How are links generated?

### DISCUSSION THREAD

**Discussion by robross0606** [pawamoy/git-changelog#95] [Q&A] [answered]
This tool is fantastic, but I'm having some difficulty with some of the automatically created links in the output.  For example, the version link at the top and the "Compare with v##.##.##" links in the changelog are blank.  I can't find anything in the documentation that references how these links are generated and it appears to be created in the template.  For example, when running against my test repo, I get this:
```
## [3.0.0]() - 2024-12-03

<small>[Compare with v2.0.0]()</small>

### Added

- Added extra mock resolvers and tests. ([50be71f](https://my.repo.com/mock/mock-service/commit/50be71f2ff86be09d7039ba59b98ee0702ac009d) by My Name).
```

Clearly it knows the URL to the commits and that link works.  However, the generated links on both "3.0.0" and "Compare with v2.0.0" are have no URL.  Why is that happening and how can it be fixed?

**Comment by pawamoy** [comment]
Thanks for the kind words!

I agree it's confusing that commits get URLs but not tags or compare links. It's because your remote URL is not recognized as a supported "provider" (GitHub, GitLab, BitBucket IIRC). When parsing commits, git-changelog sees there's no provider so it falls back to a hardcoded URL like `f"{remote_url}/commit/{hash}"`. But for tags or comparison links it doesn't fall back to hardcoded URLs. Maybe it should! If that's something you'd like, could you open a new feature request :smile:?

In the meantime, try passing `-p github` to your git-changelog invocations (or setting this in a config file).

**Comment by robross0606** [comment, marked as answer]
Thanks for the prompt reply.  This _sorta_ worked.  The result of using `--provider` was:
```
## [3.0.0](https://my.repo.com/mock/mock-service/tags/3.0.0) - 2024-12-03

<small>[Compare with v2.0.0](https://my.repo.com/mock/mock-service/compare/v2.0.0...3.0.0)</small>

### Added

- Added extra mock resolvers and tests. ([50be71f](https://my.repo.com/mock/mock-service/commit/50be71f2ff86be09d7039ba59b98ee0702ac009d) by Robert Ross).
```

The problem here is that our tag is not "3.0.0".  It is `v3.0.0".  It managed to figure out the previous tag was "v2.0.0" because "Compare with v2.0.0" references "v2.0.0".  But the URL links referencing "3.0.0" as the tag are all wrong.

**Reply by pawamoy** [reply]
Did you provide the new version when bumping, or let git-changelog guess it?

**Reply by robross0606** [reply]
I provided the new version, but I thought that had to be semver compliant so I used "3.0.0" there.  If that doesn't have to be semver complaint, I will try with "v3.0.0".

**Reply by pawamoy** [reply]
Yes please try and let me know if that worked (it should) :slightly_smiling_face:

**Reply by robross0606** [reply]
Yes, this worked.  Brilliant! 😄

**Reply by pawamoy** [reply]
Perfect, thanks for confirming!
